File: 621-task-scheduler/621-task-scheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        curTime = 0
        counts = Counter(tasks)
        heap = []
        for key,val in counts.items():
            heapq.heappush(heap,(-val,key))
        last = {}
        while heap:
            if heap[0][1] not in last:
                rem,task = heapq.heappop(heap)
                rem += 1
                if rem:
                    heapq.heappush(heap,(rem,task))
                last[task] = curTime
                curTime += 1
            else:
                temp = []
                while heap and heap[0][1] in last and (curTime-last[heap[0][1]]-1) < n:
                    temp.append(heapq.heappop(heap))

                if not heap:
                    rem,task = temp[0]
                    diff = (curTime-last[task]-1)
                    curTime += (n - diff)
                    rem += 1
                    if rem:
                        heapq.heappush(heap,(rem,task))
                    last[temp[0][1]] = curTime
                    for j in range(1,len(temp)):
                        heapq.heappush(heap,temp[j])
                else:
                    rem,task = heapq.heappop(heap)
                    rem += 1
                    if rem:
                        heapq.heappush(heap,(rem,task))
                    last[task] = curTime
                    for j in range(len(temp)):
                        heapq.heappush(heap,temp[j])
                curTime += 1
                
                
        return curTime
                
                
                
                
                
                
            
            
            

File: 621-task-scheduler/621-task-scheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        if(n==0)
            return tasks.size();
        
        int m=0;
        unordered_map<char,int> map;
        for(int i=0;i<tasks.size();i++){
            map[tasks[i]]++;
            m=max(m,map[tasks[i]]);
        }
        
        int count_max=0;
        for(auto it:map){
            if(it.second==m)
                count_max++;
        }
        
        int x = (n+1)*(m-1)+(count_max);
        int s = tasks.size();
        return max(s,x);
    }
};